File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/index/IndexDescriptor.java
/**
 * Copyright (c) 2002-2013 "Neo Technology,"
 * Network Engine for Objects in Lund AB [http://neotechnology.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.neo4j.kernel.impl.api.index;

/**
 * Description of a single index as needed by the {@link IndexProxy} cake
 *
 * This is a IndexContext cake level representation of {@link org.neo4j.kernel.impl.nioneo.store.IndexRule}
 */
public class IndexDescriptor
{
    private final long labelId;
    private final long propertyKeyId;

    public IndexDescriptor( long labelId, long propertyKeyId )
    {
        this.labelId = labelId;
        this.propertyKeyId = propertyKeyId;
    }

    @Override
    public boolean equals( Object obj )
    {
        if ( this == obj )
        {
            return true;
        }
        if ( obj != null && getClass() == obj.getClass() )
        {
            IndexDescriptor that = (IndexDescriptor) obj;
            return this.labelId == that.labelId &&
                   this.propertyKeyId == that.propertyKeyId;
        }
        return false;
    }

    @Override
    public int hashCode()
    {
        int result = (int) (labelId ^ (labelId >>> 32));
        result = 31 * result + (int) (propertyKeyId ^ (propertyKeyId >>> 32));
        return result;
    }

    public long getLabelId()
    {
        return labelId;
    }

    public long getPropertyKeyId()
    {
        return propertyKeyId;
    }
    
    @Override
    public String toString()
    {
        return String.format( "%s[label:%d, property:%d]", getClass().getSimpleName(), labelId, propertyKeyId );
    }
}

File: community/cypher/planner-spi-3.5/src/main/scala/org/neo4j/cypher/internal/planner/v3_5/spi/IndexDescriptor.scala
/*
 * Copyright (c) 2002-2019 "Neo4j,"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package org.neo4j.cypher.internal.planner.v3_5.spi

import org.neo4j.cypher.internal.planner.v3_5.spi.IndexDescriptor.{OrderCapability, ValueCapability}
import org.neo4j.cypher.internal.v3_5.logical.plans.{DoNotGetValue, GetValueFromIndexBehavior}
import org.neo4j.cypher.internal.v3_5.util.symbols.CypherType
import org.neo4j.cypher.internal.v3_5.util.{LabelId, PropertyKeyId}

sealed trait IndexLimitation
case object SlowContains extends IndexLimitation

sealed trait IndexOrderCapability {
  def asc: Boolean
  def desc: Boolean
}

object IndexOrderCapability {
  protected class BASE(override val asc: Boolean, override val desc: Boolean) extends IndexOrderCapability
  case object NONE extends BASE(false, false)
  case object ASC extends BASE(true, false)
  case object DESC extends BASE(false, true)
  case object BOTH extends BASE(true, true)
}

object IndexDescriptor {
  /**
    * Given the actual types of properties (one for a single-property index and multiple for a composite index)
    * can this index guarantee ordered retrieval?
    */
  type OrderCapability = Seq[CypherType] => IndexOrderCapability
  val noOrderCapability: OrderCapability = _ => IndexOrderCapability.NONE

  /**
    * Given the actual types of properties (one for a single-property index and multiple for a composite index)
    * does the index provide the actual values?
    */
  type ValueCapability = Seq[CypherType] => Seq[GetValueFromIndexBehavior]
  val noValueCapability: ValueCapability = s => s.map(_ => DoNotGetValue)

  def apply(label: Int, property: Int): IndexDescriptor = IndexDescriptor(LabelId(label), Seq(PropertyKeyId(property)))

  implicit def toKernelEncode(properties: Seq[PropertyKeyId]): Array[Int] = properties.map(_.id).toArray
}

case class IndexDescriptor(label: LabelId,
                           properties: Seq[PropertyKeyId],
                           limitations: Set[IndexLimitation] = Set.empty[IndexLimitation],
                           orderCapability: OrderCapability = IndexDescriptor.noOrderCapability,
                           valueCapability: ValueCapability = IndexDescriptor.noValueCapability) {
  val isComposite: Boolean = properties.length > 1

  def property: PropertyKeyId = if (isComposite) throw new IllegalArgumentException("Cannot get single property of multi-property index") else properties.head

  def canEqual(other: Any): Boolean = other.isInstanceOf[IndexDescriptor]

  // The lambda functions `orderCapability` and `valueCapability`cannot be compared for
  // equality in a sensible way. By excluding them from equals and hashCode, we make
  // the assumption that they should be always the same for (label, properties) combination
  override def equals(other: Any): Boolean = other match {
    case that: IndexDescriptor =>
      (that canEqual this) &&
        label == that.label &&
        properties == that.properties &&
        limitations == that.limitations
    case _ => false
  }

  override def hashCode(): Int = {
    val state = Seq(label, properties, limitations)
    state.map(_.hashCode()).foldLeft(0)((a, b) => 31 * a + b)
  }
}
